### PR TITLE
implement MonitorPortlet for pull-request-monitoring plugin

### PR DESCRIPTION
**Pull request by simonsymhoven**
Implement portlets for the https://github.com/jenkinsci/pull-request-monitoring-plugin.

Add one portlet for each analysis result which is added by the warnings-ng plugin to the actual run. Each portlet contains a sunburst diagram of fixed, outstanding and new issues. The new issues are clustered by severities, os it looks like this:

Bildschirmfoto 2021-05-23 um 15 53 23

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: plugin/pom.xml
Comment (KalleOlaviNiemitalo) [inline plugin/pom.xml:239]: Could this dependency be optional? The more plugins server administrators have to install, the higher the risk that one of them has a vulnerability.

Comment (simonsymhoven) [reply plugin/pom.xml:239]: Yes,  I think this can be marked as optional

**Comment by simonsymhoven** [comment]
Implement portlets for the https://github.com/jenkinsci/pull-request-monitoring-plugin.

Add one portlet for each analysis result which is added by the warnings-ng plugin to the actual run. Each portlet contains a sunburst diagram of fixed, outstanding and new issues. The new issues are clustered by severities, os it looks like this:

Bildschirmfoto 2021-05-18 um 14 33 56

**Comment by KalleOlaviNiemitalo** [comment]
This seems redundant to me because Warnings NG already has graphs for the same information.
If I understand correctly, Pull Request Monitoring only adds graphs to builds of pull requests, while Warnings NG adds them to all builds.

Perhaps this will be more useful if users can also add other graphs to the dashboard. Currently though, this MonitorPortlet implementation for Warnings NG seems to be only one available. What other portlets have been planned?

**Comment by simonsymhoven** [comment]
The idea of the plugin is to aggregate different delta metrics for pull requests and present them in a dashboard. Warning-ng is now the first plugin to provide such a portlet for this dashboard. that other developer can use the dashboard an integrate the result of the static analysis into their dashboards. The whole thing is being developed as part of a bachelor's thesis under the supervision of @uhafner. For more details I would like refer to the [Readme](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/master/README.md) or the [ux-sig](https://www.youtube.com/watch?v=F1ISpA7K0YA) meeting of 28 april.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (KalleOlaviNiemitalo) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:72]: I find the relative URL difficult to understand but it seems to work as intended.

This is called from [portlet.jelly](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/38077ebb55e65859a0e34183c4b0057764d0ea1d/src/main/resources/portlet/portlet.jelly#L25), which is referenced from [MonitoringDefaultAction/index.jelly](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/6e0e418a008f149fad739ec097bf8617ea3faff1/src/main/resources/io/jenkins/plugins/monitoring/MonitoringDefaultAction/index.jelly#L36). MonitoringDefaultAction is attached to a Run and [its getUrlName method](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/c1a90464474738e6b45d02c2272a631eae37c135/src/main/java/io/jenkins/plugins/monitoring/MonitoringDefaultAction.java#L57-L60) returns "pull-request-monitoring". So the ".." walks back to the Run and then you get to the ResultAction, which is also attached to the same Run.

Nothing to fix in this PR I think, but documenting the base URL in [MonitorPortlet.getDetailViewUrl](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/c1a90464474738e6b45d02c2272a631eae37c135/src/main/java/io/jenkins/plugins/monitoring/MonitorPortlet.java#L56-L63) might be useful, especially because the result of MonitorPortlet.getIconUrl is interpreted differently.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (KalleOlaviNiemitalo) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:52]: I wonder if this ID should also include "warnings", to reduce the risk of other plugins defining portlet IDs that clash with IDs that Warnings NG copies from AnalysisModelParserDescriptor. However, perhaps the URL would then clash anyway.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (simonsymhoven) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:72]: Thats true. It is probably easier and more understandable if I change this to an absolute url so that would be `Optional.of(action.getAbsolutUrl())`. I will expand the documentation of the method, that there is no base url, do you agree?

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (simonsymhoven) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:52]: Good point. In general it makes sense, if the artifact id of the plugin is added as a suffix.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (KalleOlaviNiemitalo) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:72]: In the old wiki, [Hyperlinks in HTML](https://wiki.jenkins.io/display/JENKINS/Hyperlinks+in+HTML) recommends not using absolute URLs that include the scheme and host. Absolute paths are OK, though. I don't know whether the advice there is still valid.

Comment (simonsymhoven) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:72]: iconUrl and detailViewUrl using both the app relative url. As soon as I released it, I will update this PR. Thanks.

Comment (uhafner) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:72]: It is still valid to always use relative paths (and never use absolute paths). Otherwise using a reverse proxy will be quite difficult. What makes more sense here: let the pr-view plugin handle the navigation from the dashboard. A portlet should only define the portion of the path that starts after the given build URL. So here:

```suggestion
        return Optional.of(action.getUrlName());
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:105]: Since you are already extending the warning plugin: move that code to the `AnalysisResult` so the code here will just be `action.getResult().hasIssues()` (tell don't ask principle).

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:113]: If the dependency is marked as optional we need to mark the extension as optional as well
```suggestion
    @Extension(optional = true)
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:118]: plural: `actions`

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:127]: This should be part of `Messages.properties` in `io.jenkins.plugins.analysis.core.portlets`.

**Review by uhafner — CHANGES_REQUESTED** [review]

**Comment by uhafner** [comment]
> Implement portlets for the https://github.com/jenkinsci/pull-request-monitoring-plugin.
> 
> Add one portlet for each analysis result which is added by the warnings-ng plugin to the actual run. Each portlet contains a sunburst diagram of fixed, outstanding and new issues. The new issues are clustered by severities, os it looks like this:
> 
> Bildschirmfoto 2021-05-18 um 14 33 56

Can you move that description to the PR description?

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/AnalysisResult.java
Comment (KalleOlaviNiemitalo) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/AnalysisResult.java:423]: Could explicitly say that this includes fixed issues.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (KalleOlaviNiemitalo) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:52]: In <https://github.com/jenkinsci/pull-request-monitoring-plugin/commit/4273a8d58824b458b3f3cf02d80664e6a23ec0d2#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R249>, you wrote

> It is recommended to use the artifact id of the plugin or parts of it as suffix.

But this is instead a prefix. Perhaps the documentation is wrong.

Comment (simonsymhoven) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:52]: Thanks. Will fix it.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet/monitor.jelly
Comment (KalleOlaviNiemitalo) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet/monitor.jelly:16]: The hasIssues check seems reversed to me, or maybe I just don't understand how j:when works.

Comment (simonsymhoven) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet/monitor.jelly:16]: Ou of course, that's stupid..

**Review by uhafner — CHANGES_REQUESTED** [review]
This is a very good starting point! 

I wonder if we could additionally add the quality gate result somewhere in the portly (if configured in the build).

**Comment by simonsymhoven** [comment]
Tank for the review. I updated the description with the latest portlet and added the quality gate (only if its not NOT_BUILD and if the result has issues).

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/AnalysisResult.java
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/AnalysisResult.java:426]:
```suggestion
        return getTotals().getTotalSize() + getTotals().getFixedSize() == 0;
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:126]: This is still a bug in my plugin. I need to replace these icons with the new Jenkins SVG status images.

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:1031]: Maybe using JSON assertions would be simpler:

```
assertThatJson(portlet.getResultIssuesAsJsonModel()).node("fixed").isEqualTo(0)
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java
Comment (simonsymhoven) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:126]: So I can leave it like this for now?

Comment (uhafner) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:126]: Yes, nothing to do on your side. I just noticed that I need to fix it in the plugin (or in Jenkins core), I just forgot about it...

**Comment by uhafner** [comment]
BTW: do not use force push after reviews. Otherwise the history in the comments is gone (and I need to recreate my local branches).

**Comment by simonsymhoven** [comment]
> BTW: do not use force push after reviews. Otherwise the history in the comments is gone (and I need to recreate my local branches).

Sorry for that. I always using a rebase to squash the bunch of commits and update the pr before merging, it's a stupid habit..

**Comment by uhafner** [comment]
> update the pr before merging

That still can be done using the other merge button.

**Comment by simonsymhoven** [comment]
> > update the pr before merging
> 
> That still can be done using the other merge button.

Thats true.

**Review by uhafner — APPROVED** [review]
Thanks, looks good now!

Just some nitpicking: can you fix the 2 PMD warnings?

One thing we should discuss (maybe before we merge this PR) if it would be better if the extension point is an abstract class. That would make it easier to change the API in the future.

**Comment by simonsymhoven** [comment]
> Thanks, looks good now!
> 
> 
> 
> Just some nitpicking: can you fix the 2 PMD warnings?
> 
> 
> 
> One thing we should discuss (maybe before we merge this PR) if it would be better if the extension point is an abstract class. That would make it easier to change the API in the future.

Oh, yes of course! 
Why should this make it easier if the factory is an abstract class?

**Comment by uhafner** [comment]
> Why should this make it easier if the factory is an abstract class? 

Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).

**Comment by KalleOlaviNiemitalo** [comment]
Do interface default methods in Java 8 change that? I guess not entirely, because interfaces still cannot define instance fields nor private methods.

**Comment by simonsymhoven** [comment]
> > Why should this make it easier if the factory is an abstract class?
> 
> Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).

So you propose to wrap both interfaces into abstract classes and add default implementations there, which other plugins have to override? But then, I can not force the user to override the methods and do not use the default implementation unless every method return null?

**Comment by uhafner** [comment]
> > > Why should this make it easier if the factory is an abstract class?
> > 
> > 
> > Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).
> 
> So you propose to wrap both interfaces into abstract classes and add default implementations there, which other plugins have to override? But then, I can not force the user to override the methods and do not use the default implementation unless every method return null?

At least the portlet seems to be something that will change from time to time. Here an abstract class (and not an interface) makes more sense. You can define all methods without any meaningful default as `abstract` now. (And some methods like default size can return something useful). The point is just in *the future* new methods can be added without breaking something. When someone implements it the first time, those `abstract` methods are just fine. 

Existing examples in Jenkins: `IssuesTotalColumn extends ListViewColumn`, `IssuesChartPortlet extends DashboardPortlet`, etc. 

I'm not sure if the same applies to the `factory`, it currently contains only the method `Collection<MonitorPortlet> getPortlets(Run<?, ?> build);`, maybe we need to update this later with a method that accepts the PR as parameter, or the target branch, etc.? Honestly, I don't know if it will be stable. But it does not hurt much to convert the interfaces to an abstract class.

**Comment by simonsymhoven** [comment]
> > > > Why should this make it easier if the factory is an abstract class?
> > > 
> > > 
> > > Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).
> > 
> > 
> > So you propose to wrap both interfaces into abstract classes and add default implementations there, which other plugins have to override? But then, I can not force the user to override the methods and do not use the default implementation unless every method return null?
> 
> At least the portlet seems to be something that will change from time to time. Here an abstract class (and not an interface) makes more sense. You can define all methods without any meaningful default as `abstract` now. (And some methods like default size can return something useful). The point is just in _the future_ new methods can be added without breaking something. When someone implements it the first time, those `abstract` methods are just fine.
> 
> Existing examples in Jenkins: `IssuesTotalColumn extends ListViewColumn`, `IssuesChartPortlet extends DashboardPortlet`, etc.
> 
> I'm not sure if the same applies to the `factory`, it currently contains only the method `Collection<MonitorPortlet> getPortlets(Run<?, ?> build);`, maybe we need to update this later with a method that accepts the PR as parameter, or the target branch, etc.? Honestly, I don't know if it will be stable. But it does not hurt much to convert the interfaces to an abstract class.

I started to wrap the interfaces into abstract classes. Will come back here and update the pr after https://github.com/jenkinsci/pull-request-monitoring-plugin/pull/75 finished and released.

**Review by uhafner — APPROVED** [review]

**Pull request merged**
